### PR TITLE
Bug 1901208: Nodes goes into NotReady state (VMware)

### PR DESCRIPTION
**Pull request by p0lyn0mial**
we will carry a patch that allows the clients (Kubelet, KCM, etc.) to detect broken connections to the api server more quickly than the default TCP timeout of 15 minutes.

**note:** this patch will be dropped in future releases once http2/ping is introduced.

the first commit uses https://github.com/openshift/library-go/pull/937

the second commit was created as follows:
`manually changed go.mod to pin to github.com/openshift/library-go v0.0.0-20201123125610-83d6d67a1e98` followed by `./hack/update-vendor.sh`

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
Both integration test (`TestCertRotation` and `TestCertRotationContinuousRequests`) seems to be permanently failing. Investigating.

File: vendor/github.com/openshift/library-go/pkg/network/dialer_linux.go [hunk 1]
```diff
@@ -16,28 +15,18 @@ func dialerWithDefaultOptions() DialContext {
 	nd := &net.Dialer{
 		// TCP_USER_TIMEOUT does affect the behaviour of connect() which is controlled by this field so we set it to the same value
 		Timeout: 25 * time.Second,
+		// KeepAlive must to be set to a negative value to stop std library from applying the default values
+		// by doing so we ensure that the options we are interested in won't be overwritten
+		KeepAlive: time.Duration(-1),
```
Comment (p0lyn0mial) [inline vendor/github.com/openshift/library-go/pkg/network/dialer_linux.go:20]: it gets overwritten here https://github.com/golang/go/blob/master/src/net/dial.go#L431

File: vendor/github.com/openshift/library-go/pkg/network/dialer_linux.go [hunk 2]
```diff
@@ -16,28 +15,18 @@ func dialerWithDefaultOptions() DialContext {
...
+		Control: func(network, address string, con syscall.RawConn) error {
+			var err error
+			err = con.Control(func(fd uintptr) {
```
Comment (p0lyn0mial) [inline vendor/github.com/openshift/library-go/pkg/network/dialer_linux.go:23]: now we are using the original file descriptor not a copy

**Comment by p0lyn0mial** [comment]
the last commit requires performing manual tests - will do that on Monday

**Comment by p0lyn0mial** [comment]
@rphillips I did manual tests this morning and haven't found any issues (https://github.com/openshift/library-go/pull/944#issuecomment-732109288). Note that I tested it against my custom applications not kubelet. 

@rphillips feel free to retitle so that it points to the BZ you want.

**Comment by rphillips** [comment]
/lgtm cancel

If we are going to carry the patch then the commit message needs to be: <carry>

**Comment by rphillips** [comment]
we need this PR to target master

**Comment by p0lyn0mial** [comment]
> we need this PR to target master

done https://github.com/openshift/kubernetes/pull/472

**Comment by nee1esh** [comment]
/retitle Bug 1901208: Nodes goes into NotReady state (VMware)

**Comment by sttts** [comment]
@cuppett for 4.5 you have to bring the change into openshift/origin.

**Pull request merged**